src/util/redux.js

```diff
@@ -0,0 +1,235 @@
+	let dispatchTree;
+
+	const reducer = createReduxReducer(reducers, initialState, rootPath);
+	const selector = selectors ? reduceSelectors(selectors) : _.identity;
+	const rootPathSelector = state => _.isEmpty(rootPath) ? state : _.get(state, rootPath);
+	const mapStateToProps = createSelector(
+		[rootPathSelector],
+		rootState => rootSelector(selector(rootState))
+	);
+	const mapDispatchToProps = dispatch => getDispatchTree(reducers, rootPath, dispatch);
+
+	return {
+		reducer,
+		connectors: [
+			mapStateToProps,
+			mapDispatchToProps,
+			mergeProps
+		]
+	};
+
+	/**
+	 * createActionCreator
+	 *
+	 * @param {function} node - a node in the the reducer tree, either a reducer or a thunk
+	 * @param {string[]} path - the path to the reducer in the reducer tree
+	 * @param {string[]} rootPath - array of strings representing the path to local state in global state
+	 * @return {function} action creator that returns either an action or a thunk
+	 */
+	function createActionCreator(node, path, rootPath) {
+		if (node.isThunk) {
+			return function thunk(...args) {
+				return function thunkInner(dispatch, getState) {
+					const localPath = _.slice(path, rootPath.length, -1);
+					const localDispatchTree = _.isEmpty(localPath) ? dispatchTree : _.get(dispatchTree, localPath);
+					const getLocalState = _.isEmpty(rootPath) ? getState : () => _.get(getState(), _.slice(path, 0, -1));
+					return node(...args)(localDispatchTree, getLocalState, dispatch, getState);
+				};
+			};
+		}
+		return function actionCreator(payload, ...meta) {
+			return {
+				type: path.join('.'),
+				payload,
+				meta
+			};
+		};
+	}
+
+	/**
+	 * createActionCreatorTree
+	 *
+	 * Walks the reducer tree and generates a tree of action creators that correspond to each reducer
+	 * @param {Object} reducers - a tree of lucid reducers
+	 * @param {string[]} rootPath - array of strings representing the path to local state in global state
+	 * @returns {Object} action creator tree
+	 */
+	function createActionCreatorTree(reducers, rootPath) {
+		return (function recur(reducers, path = []) {
+			return _.reduce(reducers, (memo, node, key) => {
+				const currentPath = path.concat(key);
+				return _.assign(memo, {
+					[key]: _.isFunction(node) ?
+						createActionCreator(node, currentPath, rootPath) :
+						recur(node, currentPath)
+				});
+			}, {});
+		})(reducers, rootPath);
+	}
+
+	/**
+	 * getDispatchTree
+	 *
+	 * Walks the reducer tree and generates an action creator tree, then binds dispatch to each node
+	 * @param {Object} reducers - a tree of lucid reducers
+	 * @param {string[]} rootPath - array of strings representing the path to local state in global state
+	 * @param {function} dispatch - the redux store's `dispatch` function
+	 */
+	function getDispatchTree(reducers, rootPath, dispatch) {
+		const actionCreatorTree = createActionCreatorTree(reducers, rootPath);
+		dispatchTree = bindActionCreatorTree(actionCreatorTree, dispatch);
```

Comment:
this is still hard because the functions that need access to `dispatchTree` are called before `dispatchTree` binding is assigned, so the reference doesn't propagate. It could be assigned to an empty "container" object and `dispatchTree` set to a property on that object, but that's pretty gross too.